### ISSUE TITLE
Quick menu freezes on macOS randomly

### ISSUE THREAD

**Issue by ppamorim** [libretro/RetroArch#13880]
# First and foremost consider this:
- Only RetroArch bugs should be filed here. Not core bugs or game bugs
- This is not a forum or a help section, this is strictly developer oriented

## Description

Sometimes the quick menu freezes when using a PS4 controller connected to the macbook pro. I need to force close the application. I wrote a script to force kill the process and open RetroArch again through my phone, but that's terrible. 

### Expected behavior

Open the quick menu.

### Actual behavior

Randomly the controls crash and I can't control the quick menu either by the PS4 or keyboard. Mouse click works but I can't move up and down.

### Steps to reproduce the bug

1. Get a Macbook Pro M1 Pro/Max and pair with a PS4 controller through bluetooth.
2. Install RetroArch Build Apr 22 2022
3. Open a game and keep playing and loading the quick menu, after a couple of minutes it should freeze.

### Bisect Results

After 20 minutes playing

### Version/Commit
Apr 22 2022

- RetroArch: GIT_VERSION (??)

### Environment information

- OS: 12.3.1 (
- Compiler: Clang 12.0.0 (from the system information)

**Comment by JasonInOttawa** [comment]
I have the same issue on a 2020 M1 MacBook Pro, running current macos etc and Retroarch 1.10.3. 

If I hit the the toggle button to go to the Quick Menu, neither my controller nor my keyboard do anything, and the mouse works a bit - I can select things with it but not scroll up or down. I have to force-quit Retroarch.

**Comment by warmenhoven** [comment]
I can't reproduce this. I think I fixed something similar to this a long time ago but can't remember clearly now. If you can still reproduce this please ping me.

**Comment by zen-logic** [comment]
This bug still exists on the current release. I am running on an M3 pro MacBook Pro

**Comment by warmenhoven** [comment]
Just to make sure I understand correctly: The bug description says "freezes", but it's not actually that the application freezes completely; it doesn't get a spinner and Finder doesn't say anything about it not being responsive; it's just that it stops responding properly to most (but not quite all) input. Correct?

Also, does it only get into this not-entirely-responsive behavior while in the menu/quick menu (as opposed to when there is no menu and the game is playing)?

**Comment by zen-logic** [comment]
That is correct. The application completely stops responding to input (either keyboard or controller). It occurs every time when opening the quick menu.

The mouse/trackpad is still somewhat useable.

I have additional information - it does *not* happen if Retroarch is running full-screen (which is a good enough workaround for me).

**Comment by warmenhoven** [comment]
I still can't reproduce this. Can you grab the latest nightly and try it again? If it's still happening, can you please give me more detailed steps (driver selections, exact button presses, etc.) for how to reproduce it?

**Comment by zen-logic** [comment]
Hi warmenhoven! I'll pull the nightly and try it when I get a minute.
In terms of reproducing this, it only happens when Video -> Full-screen Mode -> Start in Full-screen Mode is set to off.
Once  full screen mode is activated, everything works perfectly...

**Comment by zen-logic** [comment]
From my perspective, it should always be running fullscreen anyway - so maybe that should be the default behaviour?

**Comment by warmenhoven** [comment]
Ok, I've been trying to keep a close eye on this, and I've had it happen to me a couple times now. I've traced it down to the keyboard state being "stuck"; it thinks there's a key pressed when there's not. I haven't had it happen often enough to see if it's even always the same key, nor what's actually causing retroarch's key state to get stuck. I suspect it's related to window focus (e.g. another app/dialog temporarily "stealing" focus may cause this).

> Once full screen mode is activated, everything works perfectly...

Entering fullscreen causes the input driver to be reinitialized, which resets retroarch's keyboard state. So does opening/closing content (but not just the quick menu). This would end up "fixing" the problem. I don't actually have a root cause yet so can't say definitively that this does or does not happen in fullscreen mode; I suspect that it does still happen in fullscreen mode, but is harder to get into because it's harder to cause whatever sequence is causing the keyboard state to get stuck. Until I can prove one way or another, I'm going to assume the fullscreen state acts as a workaround, and is not a preventative measure.

> From my perspective, it should always be running fullscreen anyway - so maybe that should be the default behaviour?

I can see the argument for that, especially since it will cause the OS to automatically turn on Game Mode, and in fact it is the default on Steam. However I think that should be considered independent of this bug as I don't think it's actually related.

At any rate, I've just put in another version of the workaround, which is that when the app loses focus it will reset the keyboard state. Since I believe it's related to window focus, I believe that this should effectively completely cover up any problematic behavior. If anyone can reproduce this after that change, please let me know.

**Comment by zen-logic** [comment]
Hi warmenhoven!

I might be able to add something to this, which probably agrees with your previous comment.

When I encounter the problem it is always when I have launched RetroArch from ES-DE (https://es-de.org) and have never seen it if I launch RetroArch directly. It seems likely that some focus issue is probable (perhaps launching from another process/application causes focus to switch back?).

I don't have time this week, but if there's another build up with your new workaround I'll try it again at the weekend - I can replicate the issue reliably on my machine right now, so should be able to confirm if your update resolves the issue. BRB

Thanks for the great work!
